server/etcdserver/raft.go

```diff
@@ -237,6 +244,12 @@ func (r *raftNode) start(rh *raftReadyHandler) {
 				if err := r.storage.Save(rd.HardState, rd.Entries); err != nil {
 					r.lg.Fatal("failed to save Raft hard state and entries", zap.Error(err))
 				}
+
+				if walNotifyc != nil {
+					// etcdserver should wait for this notification before responding to client.
+					walNotifyc <- struct{}{}
```

Comment:
I don't think this is correct. Interestingly it's not raft code that responds to client. It's apply code.

Reply:
The exact code where client is notified is `s.w.Trigger`. For example https://github.com/etcd-io/etcd/blob/eb25e433cda952cfa2bf9ff9289b86fa8899f193/server/etcdserver/server.go#L1889